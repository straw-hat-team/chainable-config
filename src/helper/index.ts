import { stringify } from 'javascript-stringify';

export function shouldShortenFunction(value: any, verbose: boolean) {
  return typeof value === 'function' && !verbose && value.toString().length > 100;
}

export function toString(config: any, options?: { verbose: boolean }) {
  const verbose = options?.verbose ?? false;
  return stringify(config, (value, _indent, next) => {
    if (shouldShortenFunction(value, verbose)) {
      return `function () { /* omitted long function */ }`;
    }
    return next(value);
  });
}
