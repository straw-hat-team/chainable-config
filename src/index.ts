export * from './chainable';
export * from './chained-map';
export * from './chained-set';
export * from './configurable';
export * from './orderable-chained-map';
