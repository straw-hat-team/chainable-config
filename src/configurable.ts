import { toString } from './helper';

export interface ToStringOptions {
  [key: string]: any;
  verbose?: boolean;
}

export class Configurable {
  static isConfigurable(value: unknown) {
    return value instanceof Configurable;
  }

  static toConfig(value: unknown) {
    return Configurable.isConfigurable(value) ? ((value as unknown) as Configurable).toConfig() : value;
  }

  static toString(value: unknown, options?: ToStringOptions) {
    return Configurable.isConfigurable(value)
      ? ((value as unknown) as Configurable).toString(options)
      : toString(value, options);
  }

  toConfig() {
    throw new Error('toConfig method not implemented');
  }

  toString(options?: ToStringOptions) {
    const config = this.toConfig();
    return toString(config, options);
  }
}
