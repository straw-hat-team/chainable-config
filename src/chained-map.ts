import deepMerge from 'deepmerge';
import isMergeable from 'is-mergeable-object';
import { Chainable } from './chainable';
import { Configurable, ToStringOptions } from './configurable';

export type ChainedMapOptions = {
  name?: string;
  asArray?: boolean;
  emptyAsUndefined?: boolean;
};

type AliasArgs<T> = {
  key: string;
  value: T;
  alias: string;
};

export class Alias<T> extends Configurable {
  value: T;
  alias: string;
  key: string;

  constructor(args: AliasArgs<T>) {
    super();
    this.alias = args.alias;
    this.key = args.key;
    this.value = args.value;
  }

  static isAlias<T = unknown>(value: unknown): value is Alias<T> {
    return value instanceof Alias;
  }

  valueOf() {
    return {
      alias: this.alias,
      key: this.key,
      value: this.value,
    };
  }

  toConfig() {
    return Chainable.toConfig(this.value);
  }
}

export class ChainedMap<P, S = unknown> extends Chainable<P> {
  protected store = new Map<string, S | Alias<S>>();
  private options: ChainedMapOptions;

  constructor(parent: P, options: ChainedMapOptions = {}) {
    super(parent);
    this.parent = parent;
    this.options = {
      asArray: false,
      emptyAsUndefined: false,
      ...options,
    };
  }

  private computeAndSet<T extends S>(key: string, fn: () => T) {
    const value = fn();
    this.set(key, value);
  }

  static isChainedMap<P = unknown, S = unknown>(value: unknown): value is ChainedMap<P, S> {
    return value instanceof ChainedMap;
  }

  clear() {
    this.store.clear();
    return this;
  }

  delete(key: string) {
    this.store.delete(key);
    return this;
  }

  get<T extends S>(key: string): T {
    return (this.store.get(key) as unknown) as T;
  }

  set<T extends S>(key: string, value: T, options: { alias?: string } = {}) {
    const theKey = options.alias ?? key;
    const theValue = options?.alias
      ? new Alias<T>({
          key,
          value,
          alias: options.alias,
        })
      : value;

    this.store.set(theKey, theValue);
    return this;
  }

  has(key: string) {
    return this.store.has(key);
  }

  entries() {
    return this.store.entries();
  }

  keys() {
    return this.store.keys();
  }

  values() {
    return this.store.values();
  }

  getOrCompute<T extends S>(key: string, fn: () => T): T {
    if (!this.has(key)) {
      this.computeAndSet(key, fn);
    }

    return this.get<T>(key)!;
  }

  merge(values: Record<string, any>, omit: string[] = []) {
    Object.keys(values).forEach((key) => {
      if (omit.includes(key)) {
        return;
      }

      const currentValue = this.get<any>(key);
      const nextValue = values[key];

      if (ChainedMap.isChainedMap(currentValue)) {
        // @ts-ignore TODO: figure out the casting from S to ChainedMap
        currentValue!.merge(nextValue);
        return;
      }

      if (isMergeable(currentValue) && isMergeable(nextValue)) {
        const mergedValue = deepMerge<any>(currentValue!, nextValue);
        this.set(key, mergedValue);
        return;
      }

      this.set(key, nextValue);
    });

    return this;
  }

  toConfig() {
    if (this.options.emptyAsUndefined && this.store.size === 0) {
      return undefined;
    }

    if (this.options.asArray) {
      return Array.from(this.store).reduce(this.#asArrayConfig, []);
    }

    return Array.from(this.store).reduce(this.#asMapConfig, {});
  }

  toString(options?: ToStringOptions) {
    const openBracket = this.options.asArray ? '[' : '{';
    const closeBracket = this.options.asArray ? ']' : '}';
    const output: string[] = [];

    Array.from(this.store).forEach(([key, value], index, list) => {
      const isLast = index === list.length - 1;
      const name = this.options.name ?? '';
      const docKey = Alias.isAlias(value) ? value.alias : key;
      const theKey = Alias.isAlias(value) ? value.key : key;

      output.push(`/* ${name}.get(${Configurable.toString(docKey, options)}) */ `);

      if (this.options.asArray) {
        output.push(`${Configurable.toString(value, options)}`);
      } else {
        output.push(`${Configurable.toString(theKey, options)}: ${Configurable.toString(value, options)}`);
      }

      // do not add trailing comma
      if (!isLast) {
        output.push(`, `);
      }
    });

    return `${openBracket}${output.join('').trim()}${closeBracket}`;
  }

  #asArrayConfig = (config: Record<any, any>, [_key, value]: [string, S | Alias<S>]) => {
    config.push(Configurable.toConfig(value));
    return config;
  };

  #asMapConfig = (config: Record<any, any>, [key, value]: [string, S | Alias<S>]) => {
    const theKey = Alias.isAlias(value) ? value.key : key;
    config[theKey] = Configurable.toConfig(value);
    return config;
  };
}
