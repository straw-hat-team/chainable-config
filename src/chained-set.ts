import { Chainable } from './chainable';
import { Configurable } from './configurable';

export class ChainedSet<P, S = unknown> extends Chainable<P> {
  private store = new Set<S>();

  add<T extends S>(value: T) {
    this.store.add(value);
    return this;
  }

  prepend<T extends S>(value: T) {
    this.store = new Set([value, ...this.store]);
    return this;
  }

  clear() {
    this.store.clear();
    return this;
  }

  delete<T extends S>(value: T) {
    this.store.delete(value);
    return this;
  }

  values() {
    return this.store.values();
  }

  has<T extends S>(value: T) {
    return this.store.has(value);
  }

  merge<T extends S>(values: T[]) {
    this.store = new Set([...this.store, ...values]);
    return this;
  }

  toConfig() {
    return Array.from(this.values()).map(Configurable.getConfig);
  }
}
