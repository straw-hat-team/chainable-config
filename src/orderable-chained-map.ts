import { ChainedMap } from './chained-map';

export class OrderableChainedMap<P> extends ChainedMap<ChainedMap<P>> {
  private key: string;

  constructor(parent: ChainedMap<P>, key: string) {
    super(parent);
    this.key = key;
  }

  before(key: string) {
    return this.parent.move(this.key, ({ before }) => before(key));
  }

  after(key: string) {
    return this.parent.move(this.key, ({ after }) => after(key));
  }
}
