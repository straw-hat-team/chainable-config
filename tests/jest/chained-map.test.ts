import { ChainedSet } from '../../src/chained-set';
import { ChainedMap } from '../../src/chained-map';

it('validates if a value is a chainable map object', () => {
  const chainMap = new ChainedMap(undefined);
  expect(ChainedMap.isChainedMap(chainMap)).toBeTruthy();
  expect(ChainedMap.isChainedMap('helloi')).toBeFalsy();
});

describe('Given a ChainedMap object', () => {
  describe('When calling .set()', () => {
    it('Then saves the value', () => {
      const chaiMap = new ChainedMap(undefined);
      const context = chaiMap.set('name', 'pepega');
      expect(chaiMap.has('name')).toBeTruthy();
      expect(context).toBe(chaiMap);
    });
  });

  describe('When calling .delete()', () => {
    it('Then removes the value', () => {
      const chaiMap = new ChainedMap(undefined);
      chaiMap.merge({ name: 'pepega', reaction: 'pogchamp' });
      const context = chaiMap.delete('name');
      expect(chaiMap.has('name')).toBeFalsy();
      expect(context).toBe(chaiMap);
    });
  });

  describe('When calling .keys()', () => {
    it('Then returns the list of keys', () => {
      const chaiMap = new ChainedMap(undefined);
      chaiMap.merge({ name: 'pepega', reaction: 'pogchamp' });
      const keys = Array.from(chaiMap.keys());
      expect(keys).toEqual(['name', 'reaction']);
    });
  });

  describe('When calling .clear()', () => {
    it('Then removes all the values', () => {
      const chaiMap = new ChainedMap(undefined);
      chaiMap.merge({ name: 'pepega', reaction: 'pogchamp' });
      const context = chaiMap.clear();
      const entries = Array.from(chaiMap.entries());
      expect(entries).toEqual([]);
      expect(context).toBe(chaiMap);
    });
  });

  describe('When calling .values()', () => {
    it('Then returns the list of values', () => {
      const chaiMap = new ChainedMap(undefined);
      chaiMap.merge({
        name: 'pepega',
        reaction: 'pogchamp',
        channel: 'alchemist_ubi',
      });
      const values = Array.from(chaiMap.values());
      expect(values).toEqual(['pepega', 'pogchamp', 'alchemist_ubi']);
    });
  });

  describe('When calling .entries()', () => {
    describe('When have entries', () => {
      it('Then returns the list of entries', () => {
        const chaiMap = new ChainedMap(undefined);
        chaiMap.merge({
          name: 'pepega',
          reaction: 'pogchamp',
          channel: 'alchemist_ubi',
        });
        const entries = Array.from(chaiMap.entries());
        expect(entries).toEqual([
          ['name', 'pepega'],
          ['reaction', 'pogchamp'],
          ['channel', 'alchemist_ubi'],
        ]);
      });
    });
  });

  describe('When calling .has()', () => {
    describe('When the value does not exists', () => {
      it('Then returns false', () => {
        const chaiMap = new ChainedMap(undefined);
        expect(chaiMap.has('name')).toBeFalsy();
      });
    });
    describe('When the value does exists', () => {
      it('Then returns true', () => {
        const chaiMap = new ChainedMap(undefined);
        chaiMap.set('name', 'pepega');
        expect(chaiMap.has('name')).toBeTruthy();
      });
    });
  });

  describe('When calling .merge()', () => {
    it('Then merges the values', () => {
      const chaiMap = new ChainedMap(undefined);
      chaiMap.merge({
        name: 'pepeg',
        hello: 'world',
        items: [1, 2, 3],
      });
      const context = chaiMap.merge({
        newKey: true,
        name: 'pepega',
        items: [4, 5, 6],
      });
      const entries = Array.from(chaiMap.entries());
      expect(entries).toEqual([
        ['name', 'pepega'],
        ['hello', 'world'],
        ['items', [1, 2, 3, 4, 5, 6]],
        ['newKey', true],
      ]);
      expect(context).toBe(chaiMap);
    });

    describe('When omitting keys', () => {
      it('Then merges the values omitting the keys', () => {
        const chaiMap = new ChainedMap(undefined);
        const context = chaiMap.merge(
          {
            name: 'pepeg',
            hello: 'world',
            items: [1, 2, 3],
          },
          ['items']
        );
        const entries = Array.from(chaiMap.entries());
        expect(entries).toEqual([
          ['name', 'pepeg'],
          ['hello', 'world'],
        ]);
        expect(context).toBe(chaiMap);
      });
    });
  });

  describe('When calling .getOrCompute()', () => {
    describe('When the key is not present', () => {
      it('Then sets and returns the value', () => {
        const setter = () => 'alpha';
        const chaiMap = new ChainedMap(undefined);
        expect(chaiMap.getOrCompute('a', setter)).toBe('alpha');
        expect(chaiMap.get('a')).toBe('alpha');
      });
    });

    describe('When the key is present', () => {
      it('Then returns the value', () => {
        const setter = () => 'alpha';
        const chaiMap = new ChainedMap(undefined);
        chaiMap.set('a', 1);
        expect(chaiMap.getOrCompute('a', setter)).toBe(1);
        expect(chaiMap.get('a')).toBe(1);
      });
    });
  });

  describe('When calling .toConfig()', () => {
    describe('When does not have entries', () => {
      it('Then returns a empty configuration object', () => {
        const chaiMap = new ChainedMap(undefined);
        const config = chaiMap.toConfig();
        expect(config).toEqual({});
      });
    });
    describe('When have some entries', () => {
      it('Then returns a configuration object', () => {
        const chaiMap = new ChainedMap(undefined);
        const chaiMap2 = new ChainedMap(undefined);
        chaiMap2.merge({
          name: 'pogchamp',
          hello: 'ubi',
        });
        chaiMap.merge({
          name: 'pepeg',
          hello: 'world',
          answer: chaiMap2,
        });
        const config = chaiMap.toConfig();
        expect(config).toEqual({
          name: 'pepeg',
          hello: 'world',
          answer: {
            name: 'pogchamp',
            hello: 'ubi',
          },
        });
      });
    });
  });

  describe('When asArray configuration is set', () => {
    it('Then returns the config as an array', () => {
      const chaiMap = new ChainedMap(undefined, { asArray: true });
      const complex = new ChainedSet(undefined).add('nice');
      const config = chaiMap.set('js', 'js').set('jsx', 'jsx').set('complex', 'jsx').set('complex', complex).toConfig();

      expect(config).toEqual(['js', 'jsx', ['nice']]);
    });
  });
});
