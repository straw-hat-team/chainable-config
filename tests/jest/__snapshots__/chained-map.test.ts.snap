// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a ChainedMap object When calling .toString() Then returns the correct string 1`] = `"['js','jsx',['nice']]"`;
